### PR TITLE
Fix #3015: exhaustivity check on top of native apply

### LINKED ISSUE #3015: Remove magic in exhaustivity check
In the exhaustivity check, when decomposing a type `path.B[Int, String]`, we get two child classes `C1` and `C1`. The problem is how to turn the two classes into proper types related to `path`, `Int` and `String`?

Current implementation resorts to some ad-hoc hack which is not well-grounded. The correct way to solve the problem is to use type inference, as suggested by Martin.

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #3015: exhaustivity check on top of native apply

### CODE REVIEW

**Comment by liufengyun** [comment]
test performance please

**Comment by liufengyun** [comment]
The [performance graph](http://dotty-bench.epfl.ch/3074) shows that the new exhaustivity check performs the same on normal patmat tests as before, but degrades significantly on the abnormal test case [V](https://github.com/liufengyun/bench/blob/master/tests/exhaustivity-V.scala): from 2s to 3.5s.

**Comment by odersky** [comment]
We see now a much bigger performance gain than with applied types alone on dotc. It's about 10% faster as opposed to maybe 2 or 3% before. Interesting, given that exhaustivity checks are now slower than before. My guess is that the reporting of the spurious exhaustivity warnings in dotc itself (there were about 300 of them) impacted overall performance. Anyway, getting 10% on dotc and 6-7 % on Vector or stdlib is very encouraging!

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 1]
```diff
@@ -403,20 +403,31 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
...
-    case Typed(expr, _) => Typ(expr.tpe.stripAnnots, true)
+    case Typed(expr, _) => Typ(erase(expr.tpe.stripAnnots), true)
     case _ =>
       debug.println(s"unknown pattern: $pat")
       Empty
   }
 
+  /* Erase a type binding according to erasure semantics in pattern matching */
+  def erase(tp: Type): Type = tp match {
+    case tp@AppliedType(tycon, args) => erase(tp.superType)
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Useless erase, this is thrown away

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 2]
```diff
@@ -403,20 +403,31 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
...
+      if (tycon.isRef(defn.ArrayClass)) tp.derivedAppliedType(tycon, args.map(erase))
+      else tp.derivedAppliedType(tycon, args.map(t => WildcardType(TypeBounds.empty)))
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: I think you can just use the WildcardType object here.

**Comment by liufengyun** [comment]
@odersky The printing of exhaustivity warnings involves a lot of IO, which has a bad impact on performance. You can also test performance by disable the exhautivity check in the pattern matcher, and see how it improves the numbers.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 3]
```diff
@@ -598,10 +579,34 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
     val tvars = tp1.typeParams.map { tparam => newTypeVar(tparam.paramInfo.bounds) }
     val protoTp1 = thisTypeMap(tp1.appliedTo(tvars))
 
-    if (protoTp1 <:< tp2 && isFullyDefined(protoTp1, ForceDegree.all)) protoTp1
+    // replace type parameter references with fresh type vars or bounds
+    val typeParamMap = new TypeMap {
+      def apply(t: Type): Type = t match {
+
+        case tp: TypeRef if tp.underlying.isInstanceOf[TypeBounds] =>
+          // See tests/patmat/gadt.scala  tests/patmat/exhausting.scala
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Add: if tp.symbol.is(TypeParam) &&

**Comment by smarter** [comment]
> The performance graph shows that the new exhaustivity check performs the same on normal patmat tests as before, but degrades significantly on the abnormal test case V: from 2s to 3.5s.

Do you know why https://github.com/liufengyun/bench/blob/master/tests/exhaustivity-V.scala performs worse now?

**Comment by liufengyun** [comment]
@smarter I think it's because the type inference approach is more expensive than the previous hack for generics.

**Comment by smarter** [comment]
But there is nothing to infer in https://github.com/liufengyun/bench/blob/master/tests/exhaustivity-V.scala, right? Maybe we could avoid the expensive use of type inference in some cases. (Of course this shouldn't be required to get this PR in, but maybe something to investigate later)

**Comment by liufengyun** [comment]
@smarter Good point, as `Product` is not decomposable, the code path should be the same as before for the test `V`, as we didn't make any changes to the space algorithm.

**Comment by liufengyun** [comment]
Superseded by #3061

**Pull request closed without merging**